feat: 회원 API  기능 추가

shindawull opened the pull request:
Reverts
사유는 aop 미적용으로 인해 추가 개발 후 다시 코드리뷰를 위한 revert

shindawull committed:
feat: (issue/2) 회원 api 개발
- 재가입 조건 추가 (조건:탈퇴날짜 기준으로 30일 이후 재가입 가능)
- 사용안하는 메서드파라미터 제거
- 기능에 맞게 메서드명 수정
- 로그레벨에 맞춰 로그 수정

shindawull committed:
feat: (issue/2) 회원 api 개발
- pom.xml에 aop dependency 추가
- 로그인 AOP 추가 및 적용
- SessionUtils 추가 및 관련된 조건문, 변수 수정
- enum 다중조건문을 람다식으로 수정
- 회원 수정 시 유저타입 조건문 제거

shindawull committed:
feat: (issue/2) 회원 api 개발
- 공통 Response 객체 추가, 적용
- Exception 추가, 적용

ccommit-dev commented on src/main/java/com/ccommit/fashionserver/aop/CommonResponse.java:
```diff
@@ -0,0 +1,28 @@
+ */
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
```
@shindawull 안녕하세요 다울님
@NoArgsConstructor, @AllArgsConstructor 두어노테이션의 기능을 
각각 어노테이션 위에 주석으로 추가 부탁드립니다.

shindawull replied:
네

ccommit-dev commented on src/main/java/com/ccommit/fashionserver/aop/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,51 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(value = NullPointerException.class)
```
@shindawull 다울님 사용자 지정 예외를 공통으로 처리하기 위해서 
GlobalExceptionHandler 를 구현하신거로 보입니다. 다만, 
NullPointerException, DuplicateException, NotMatchException, PermissionDeniedException 
4가지 예외를 구현할떄마다 GlobalExceptionHandler 에 추가하는것은 추상화되지 않고
계속 추가해줘야 하기 때문에 RuntimeException이나 FashionServerExcepion으로 추상화 하셔야합니다.
수정 부탁드립니다.

shindawull replied:
네 반영하였습니다.

ccommit-dev commented on src/main/java/com/ccommit/fashionserver/aop/LoginCheckAspect.java:
```diff
@@ -0,0 +1,70 @@
+        int index = 0;
+
+        for (int i = 0; i < loginCheck.types().length; i++) {
+            if (id == 0) {
```
id값은 항상 0 인데   if (id == 0) { 코드는 불필요한 코드로 보입니다. 수정 부탁드립니다.

shindawull replied:
네 반영하였습니다.

ccommit-dev commented on src/main/java/com/ccommit/fashionserver/controller/UserController.java:
```diff
@@ -34,55 +39,74 @@
 public class UserController {
     @Autowired
-    private UserService userService;
+    private final UserService userService;
```
final 키워드를 사용하신 이유를 댓글로 작성 부탁드립니다.

shindawull replied:
의존성 주입 시 객체가 변하지 않게 하기 위해서 사용하였습니다.

ccommit-dev commented on src/main/java/com/ccommit/fashionserver/controller/UserController.java:
```diff
@@ -34,55 +39,74 @@
-        return result;
+            throw new NotMatchException("회원정보가 불일치합니다. 확인해주세요.");
+
+        CommonResponse<String> response = new CommonResponse<>(HttpStatus.OK, "SUCCESS", "정상적으로 회원정보가 수정되었습니다.", null);
```
CommonResponse 생성 body에 userDto를 넣어주어 연동한 프론트 서버에게 정보를 주는것이 좋을것 같습니다.

shindawull replied:
네 반영하였습니다.

ccommit-dev commented on src/main/java/com/ccommit/fashionserver/service/UserService.java:
```diff
@@ -39,40 +48,44 @@ public boolean isExistId(String userId) {
+                            throw new NullPointerException("존재하지 않는 회원 타입입니다.");
+                        }
+                    });
+            userMapper.signUp(userDto);
```
userMapper.signUp()  메서드 에서 예외가 발생하면 현재 어떻게 처리되는지 댓글로 작성 부탁드립니다.

ccommit-dev commented on src/main/resources/mapper/userMapper.xml:
```diff
@@ -21,11 +22,23 @@
+    <select id="isJoinPossible" resultType="int">
+        SELECT COUNT(A.id)
+        FROM users as A,
+            (SELECT id, user_id, DATE_ADD(DATE_FORMAT(update_date, '%Y-%m-%d'), INTERVAL 30 DAY) as joinPossibleDate ,
```
서브쿼리 SELECT 문보다는 탈퇴한지 30일이 지난 유저 정보만 전체 가져와서 Service로직에서 가입하려는 유저의 정보와 일치하는게 있는지 확인하는게 성능상으로는 더  좋을것 같습니다. 참고 부탁드립니다.

shindawull replied:
@shindawull 서브쿼리가 성능에 왜 안좋은지 서치하여 댓글로 추가 정보 달겠습니다.

shindawull replied:
서브쿼리 사용 시 있을 문제.
서브쿼리는 서브쿼리를 접근할 때마다 SELECT구문을 실행해서 데이터를 만들어내지만 값을 저장하는 건 아니기 때문에 데이터 INPUT OUTPUT 비용이 발생합니다. 
그리고 여러 계층에 걸쳐 SELECT구문이 있기때문에 가독성이 떨어질 수 있습니다.

shindawull committed:
feat: (issue/2) 회원 api 개발
- 로그인 AOP 코드 수정
- ExceptionHandler 수정
- Custom Exception 추가
- Custom Error Code 추가
- Annotation 기능설명 주석 추가

shindawull committed:
feat: (issue/2) 회원 api 개발
- 회원가입 시 탈퇴조건 수정
- 서브쿼리 제거
- 가입가능한 날짜 구하는 쿼리 추가
